python/sglang/srt/managers/router/model_rpc.py

```diff
@@ -463,15 +475,28 @@ def forward_decode_batch(self, batch: Batch):
         batch.prepare_for_decode()
 
         # Forward
-        logits = self.model_runner.forward(batch, ForwardMode.DECODE)
-        next_token_ids, next_token_probs = batch.sample(logits)
+        logits, (logprobs, _) = self.model_runner.forward(
+            batch,
+            ForwardMode.DECODE,
+            batch.return_logprob,
+        )
+        next_token_ids, _ = batch.sample(logits)
         next_token_ids = next_token_ids.cpu().tolist()
 
         # Check finish condition
         reqs = batch.reqs
-        for i in range(len(reqs)):
-            reqs[i].output_ids.append(next_token_ids[i])
-            reqs[i].check_finished()
+        for i, (req, next_tok_id) in enumerate(zip(reqs, next_token_ids)):
+            req.output_ids.append(next_tok_id)
+            req.check_finished()
+
+            if logprobs is not None:
+                next_tok_text = req.tokenizer.convert_ids_to_tokens([next_tok_id])[0]
+                next_tok_text = (
+                    next_tok_text.decode() if isinstance(next_tok_text, bytes) else next_tok_text
+                )
+                req.token_logprob.append(
+                    (next_tok_text, next_tok_id, logprobs[i][next_tok_id].tolist())
```

Comment:
The index operation can be batched, so we only move tensors from GPU to CPU once.

```python
selected_logprobs = logprobs[torch.arange(len(reqs)), next_token_ids].tolist()
...
req.token_logprob.append(
  (next_tok_text, next_tok_id, logprobs[i][next_tok_id].tolist())
```